src/solvers/dgsem_structured/dg.jl

```diff
@@ -22,9 +22,10 @@ function create_cache(mesh::StructuredMesh, equations::AbstractEquations, dg::DG
 end
 
 # Extract contravariant vector Ja^i (i = index) as SVector
+# TODO: Here, size() causes a lot of allocations! Find an alternative to improve performance
 @inline function get_contravariant_vector(index, contravariant_vectors, indices...)
     SVector(ntuple(@inline(dim->contravariant_vectors[dim, index, indices...]),
-                   Val(ndims(contravariant_vectors) - 3)))
+                   Val(size(contravariant_vectors, 1))))
 end
```

Comment:
This new version of `get_contravariant_vector` causes too many allocations. Unfortunately, `ndims(contravariant_vectors) - 3` does not work for the simulations on the spherical shell and `size()` is evaluated in runtime. Is there a better way to do this?

Reply:
Is this code only called from your code or also from existing routines? In the current implementation it is not type stable, thus the allocations.

We might get around it easily if it is only used from code you control, and harder (but not impossible) if it is in generic routines as well.

Reply:
No, this is the existing volume and surface integral routines. In fact, I temporarily changed the existing volume integral calls, such that the code runs a bit faster. See, e.g., 
https://github.com/trixi-framework/Trixi.jl/blob/b89e333af4fd1b9c84a3413511c6939e4f7d8458/src/solvers/dgsem_structured/dg_2d.jl#L78

Reply:
Hm. I wonder if what you are doing in this PR is fundamental enough of a conceptual difference to require a new mesh type. That is, to not try to shoehorn this into `P4estMesh{2}` but rather something like `P4estMesh{3/2}`?

Reply:
Alternatively, we may need to pass an additional argument to `get_contravariant_vector` that determines the number of dimensions at compile time - the `mesh`?

Reply:
True. The question is whether the mesh truly "owns" the dimensionality of a problem. But it would certainly be less intrusive than adding a new mesh type.

Reply:
Well, it looks like there isn't anything like "the dimensionality of a problem" anymore. We need to distinguish between the dimensionality of the geometry and the equations. From my point of view, the contravariant vectors are clearly associated with the geometry, i.e., the `mesh`.

However, this is really a tough problem since we want to pass the contravariant vectors as normal directions to the `flux`es. Thus, we assume that the dimensionality of the `equations` and the `mesh` are the same. This requires some really careful design to achieve a consistent splitting.

Reply:
> We need to distinguish between the dimensionality of the geometry and the equations.

Exactly!
The contravariant vectors are actually associated with both the dimensionality of the geometry and the equations. For instance:
* For a 2D geometry and a 2D equation, the contravariant vectors are of size `2 x 2 x nnodes x nnodes x nelements`.
* For a 2D geometry and a 3D equation, the contravariant vectors are of size `3 x 3 x nnodes x nnodes x nelements`.
* For a 3D geometry and a 3D equation, the contravariant vectors are of size `3 x 3 x nnodes x nnodes x nnodes x nelements`.

For this particular function, we would actually need to pass the number of dimensions of the equation to define the size of the SVector... Or maybe there is a workaround. :thinking:

Reply:
> Hm. I wonder if what you are doing in this PR is fundamental enough of a conceptual difference to require a new mesh type. That is, to not try to shoehorn this into P4estMesh{2} but rather something like P4estMesh{3/2}?

This function is mesh type agnostic, so that won't fix this particular problem.